### ISSUE TITLE
Direct-to-BIO SSL mode

### ISSUE THREAD

**Issue by indutny** [openssl/openssl#1333]
At the moment, all OpenSSL versions will first read/write record to the internal buffer, and only later call `BIO_write` with this pointer.

This is absolutely fine for `BIO_s_connect()`/`BIO_s_accept()`. However for memory BIOs (including custom BIOs that we use in Node.js) set through `SSL_set_bio`, this performs slower because of the extra `memcpy` call that needs to be done.

In case of Node.js, our custom BIO is powered by a circular buffer, and it is pretty easy to pre-allocate enough data to fit the record.

Hereby, I propose to add APIs for OpenSSL to introduce Direct-to-BIO operation mode in which special ctrl methods will be called on BIO, and if it recognizes them - the record data will be read/written directly to the BIO's internal buffer.

Does this sound reasonable?

cc @richsalz

**Comment by richsalz** [comment]
It’s a good goal, not sure about the best way to implement it.  Ok? :)

**Comment by indutny** [comment]
Yep, thank you for response!

As far as I got it there has been some recent changes with the way write buffers are managed. There seem to be multiple pipelines for them now.

cc @mattcaswell since you implemented this.

**Comment by indutny** [comment]
The APIs may look like this:

* `BIO_CTRL_GET_FRAGMENT_BUFFER` - to get a pointer to the bio's internal buffer
* `BIO_CTRL_COMMIT_FRAGMENT` - to commit fragment to BIO

**Comment by mattcaswell** [comment]
Yes, this could be possible. Probably not a top priority though. I'll put this against the post 1.1.0 milestone.

**Comment by indutny** [comment]
Thank you!

**Comment by Drawaes** [comment]
Hi,
I would just like to say I am writing an openssl channel on the new .net channels. This would be useful as the entire stack is close to no copy except for this one tiny part. I have a custom bio pointing to my own buffers, and this would work perfectly. 
https://github.com/davidfowl/Channels/pull/128
Thanks

**Comment by kaduk** [comment]
In the absence of a contributed pull request, this seems unlikely to happen for 1.1.1, so I'll mark it as post-1.1.1 for now.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input
